src/utils/rTree.ts

```diff
@@ -0,0 +1,363 @@
+/**
+ * Copyright 2017-present Palantir Technologies
+ * @license MIT
+ */
+
+import { Bounds, IEntityBounds, Point } from "../core/interfaces";
+import {
+    IRTreeSplitStrategy,
+    NodePair,
+    SplitStrategyLinear,
+} from "./rTreeSplitStrategies";
+
+/**
+ * The maximum number of children in an r-tree node before we attempt to split.
+ * This must be >= 2.
+ */
+const DEFAULT_MAX_NODE_CHILDREN = 5;
+
+/**
+ * There are several strategies for splitting nodes that contain overlapping
+ * regions. By default we use `SplitStrategyLinear` which minimizes the change
+ * in node bounding box area.
+ */
+const DEFAULT_SPLIT_STRATEGY: IRTreeSplitStrategy = new SplitStrategyLinear();
+
+/**
+ * R-Tree is a multidimensional spatial region tree. It stores entries that have
+ * arbitrarily overlapping bounding boxes and supports efficient point and
```

Comment:
nit: can we document runtime and space complexity to be more precise with what "efficient" means